builtin/mainmenu/dlg_mg_flags.lua

```diff
@@ -0,0 +1,243 @@
+			first = false
+		end
+		return flag_str
+	end
+	
+	core.setting_set("mg_flags", mg_flags_helper(flag_list_g))
+	core.setting_set("mgv6_spflags", mg_flags_helper(flag_list_6))
+	core.setting_set("mgv7_spflags", mg_flags_helper(flag_list_7))
+	core.setting_set("mgflat_spflags", mg_flags_helper(flag_list_f))
+	core.setting_set("mg_valleys_spflags", mg_flags_helper(flag_list_v))
+	core.setting_set("water_level", water_level)
+end
+
+-- Load default settings if not exist
+if core.setting_get("water_level") == nil then
+	core.setting_set("water_level", 1)
+end
+if core.setting_get("mg_flags") == nil then
+	core.setting_set("mg_flags", "caves, dungeons, light, decorations")
+end
+if core.setting_get("mgv6_spflags") == nil then
+	core.setting_set("mgv6_spflags", "jungles, biomeblend, mudflow, snowbiomes, trees")
+end
+if core.setting_get("mgv7_spflags") == nil then
+	core.setting_set("mgv7_spflags", "mountains, ridges")
+end
+if core.setting_get("mgflat_spflags") == nil then
+	-- prevents error because of nil value
+	core.setting_set("mgflat_spflags", "")
+end
+if core.setting_get("mg_valleys_spflags") == nil then
+	core.setting_set("mg_valleys_spflags", "altitude_chill, humid_rivers")
+end
+
+local function load_mg_flags()
+	local function mg_flags_helper(flag_list,text)
+		local text = string.split(text, ", ")
+		for i = 1, #text do
+			if text[i]:sub(1, 2) == "no" then
+				flag_list[text[i]:sub(3)] = false
+			else
+				flag_list[text[i]] = true
+			end
+		end
+	end
+	mg_flags_helper(flag_list_g, core.setting_get("mg_flags"))
+	mg_flags_helper(flag_list_6, core.setting_get("mgv6_spflags"))
+	mg_flags_helper(flag_list_7, core.setting_get("mgv7_spflags"))
+	mg_flags_helper(flag_list_f, core.setting_get("mgflat_spflags"))
+	mg_flags_helper(flag_list_v, core.setting_get("mg_valleys_spflags"))
+	water_level = core.setting_get("water_level")
+	
+	update_mg_flags()
+end
+
+local function mg_flags_formspec(dialogdata)
+	-- Load minetest.conf for first time
+	if mg_first_time == true then
+		load_mg_flags()
+		update_mg_flags()
+	end
+	mg_first_time = false
+	
+	local mgname = "[" .. core.setting_get("mg_name") .. "]"
+	local formspec =
+		"size[8,5,true]" ..
+		"button[5.5,4.6;2.5,0.5;mg_close;" .. fgettext("Close") .. "]" ..
+		
+	-- Special mapgen-specific flags
+		"label[0,0;" .. minetest.formspec_escape(mgname) .. " " .. fgettext("Mapgen flags") .. "]"
+	if core.setting_get("mg_name") == "v5" then
+		formspec = formspec ..
+			"label[0,1.0;" .. fgettext("No flags in v5 mapgen") .. "]"
+	elseif core.setting_get("mg_name") == "v6" then
+		-- If snowbiomes enabled, tell player if jungle is auto-enabled, too.
+		if flag_list_6["snowbiomes"] == true then
+			flag_list_6["jungles"] = true  -- Do not make the player get confused
+			formspec = formspec ..
+			"label[0.7,2.2;" .. fgettext("(with Jungles)") .. "]"
+		else
+			formspec = formspec ..
+			"checkbox[0.25,2.0;cb_jungles;" .. fgettext("Jungles") .. ";" ..
+				dump(flag_list_6["jungles"]) .. "]"
```

Comment:
The output of `dump` is not garantueed to have a fixed format. Please use `tostring` or a custom function, if needed here.